locator/topology.hh

```diff
@@ -24,24 +24,113 @@ using namespace seastar;
 
 namespace locator {
 
+class topology;
+
+class node : public enable_lw_shared_from_this<node> {
+public:
+    using local = bool_class<struct local_tag>;
+
+private:
+    host_id _host_id;
+    inet_address _endpoint;
+    endpoint_dc_rack _dc_rack;
+
+    // Private state of this instance.
+    // May change across topology generations
+    local _is_local;
+
+    friend class topology;
+public:
+    node(host_id id, inet_address endpoint, endpoint_dc_rack dc_rack, local is_local);
+
+    node(const node&) = default;
+    node(node&&) = delete;
+
+    const host_id& host_id() const noexcept {
+        return _host_id;
+    }
+
+    const inet_address& endpoint() const noexcept {
+        return _endpoint;
+    }
+
+    const endpoint_dc_rack& dc_rack() const noexcept {
+        return _dc_rack;
+    }
+
+    local is_local() const noexcept { return _is_local; }
+};
+
+using node_ptr = lw_shared_ptr<const node>;
+using node_set = std::unordered_set<node_ptr>;
+
 class topology {
 public:
     struct config {
+        host_id local_host_id;
+        inet_address local_endpoint;
         endpoint_dc_rack local_dc_rack;
         bool disable_proximity_sorting = false;
     };
     topology(config cfg);
-    topology(topology&&) = default;
+    topology(topology&&) noexcept;
 
     topology& operator=(topology&&) = default;
 
     future<topology> clone_gently() const;
     future<> clear_gently() noexcept;
 
+public:
+    const node_ptr& local_node() const noexcept {
+        return _local_node;
+    }
+
+    // Adds a node with given host_id, endpoint, and DC/rack.
+    node_ptr add_node(host_id id, const inet_address& ep, const endpoint_dc_rack& dr);
+
+    // Optionally updates node's current host_id, endpoint, or DC/rack.
+    node_ptr update_node(node_ptr node, std::optional<host_id> opt_id, std::optional<inet_address> opt_ep, std::optional<endpoint_dc_rack> opt_dr);
```

Comment:
How can we change a node's host id? Isn't host id the identity of the node? Or is this to support some current "legacy" use-cases that are going away?

Reply:
There was a legacy path where the replacing node that inherited the replaced node's host-id would start up with a random host id and later update it to the replaced node's id.

I'll see if we can indeed eliminate this after https://github.com/scylladb/scylladb/pull/12250

Reply:
Also, https://github.com/scylladb/scylladb/blob/64102780fe570b53617594900eba1e229e537a3c/locator/topology.cc#L34
creates the local node initially with no host_id.

In the context of this branch, this translated into https://github.com/scylladb/scylladb/blob/32ba293bf9f9f56bc10b5258f78c1f88a52887ff/locator/topology.cc#L60

And later on the host_id is "patched in", in https://github.com/scylladb/scylladb/blob/32ba293bf9f9f56bc10b5258f78c1f88a52887ff/main.cc#L1172

Reply:
Didn't we kill the replace node that takes over the id?

If so, can make nodes always have ids? This violation of a basic invariant is bad.